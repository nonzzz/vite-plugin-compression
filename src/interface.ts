import type { FilterPattern } from '@rollup/pluginutils'
import type { HookHandler, Plugin } from 'vite'
import type { BrotliOptions, InputType, ZlibOptions, ZstdOptions } from 'zlib'

export type CoreAlgorithm = 'gzip' | 'brotliCompress' | 'deflate' | 'deflateRaw' | 'zstandard'

export type AliasAlgorithm = 'gz' | 'br' | 'brotli' | 'zstd'

export type Algorithm = CoreAlgorithm | AliasAlgorithm
export interface UserCompressionOptions {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  [key: string]: any
}

export type InferDefault<T> = T extends infer K ? K : UserCompressionOptions

export type CompressionOptions<T> = InferDefault<T>

export type Pretty<T> =
  & {
    [key in keyof T]: T[key]
  }
  & NonNullable<unknown>

export interface FileNameFunctionMetadata {
  // eslint-disable-next-line no-use-before-define
  algorithm: CoreAlgorithm | AlgorithmFunction<UserCompressionOptions>
  options: UserCompressionOptions
}

export type LogLevel = 'info' | 'silent'

interface BaseCompressionPluginOptions {
  include?: FilterPattern
  exclude?: FilterPattern
  threshold?: number
  filename?: string | ((id: string, metadata: FileNameFunctionMetadata) => string)
  deleteOriginalAssets?: boolean
  skipIfLargerOrEqual?: boolean
  logLevel?: LogLevel
}
export interface AlgorithmToZlib {
  gz: ZlibOptions
  gzip: ZlibOptions
  brotliCompress: BrotliOptions
  brotli: BrotliOptions
  br: BrotliOptions
  deflate: ZlibOptions
  deflateRaw: ZlibOptions
  zstd: ZstdOptions
  zstandard: ZstdOptions
}

export type AlgorithmFunction<T extends UserCompressionOptions> = (buf: InputType, options: T) => Promise<Buffer>

export type defineAliasAlgorithmResult<T extends UserCompressionOptions = UserCompressionOptions> =
  | readonly [
    'gz',
    ZlibOptions
  ]
  | readonly [
    'br' | 'brotli',
    BrotliOptions
  ]
  | readonly [
    'zstd',
    ZstdOptions
  ]
  | readonly [
    AlgorithmFunction<T>,
    T
  ]

export type DefineAlgorithmResult<T extends UserCompressionOptions = UserCompressionOptions> =
  | readonly [
    'gzip' | 'deflate' | 'deflateRaw',
    ZlibOptions
  ]
  | readonly [
    'brotliCompress',
    BrotliOptions
  ]
  | readonly [
    'zstandard',
    ZstdOptions
  ]
  | readonly [
    AlgorithmFunction<T>,
    T
  ]

export type Algorithms = (Algorithm | DefineAlgorithmResult | defineAliasAlgorithmResult)[]

export interface ViteCompressionPluginOption extends BaseCompressionPluginOptions {
  algorithms?: Algorithms
}

export type GenerateBundle = HookHandler<Plugin['generateBundle']>

export type WriteBundle = HookHandler<Plugin['writeBundle']>

export interface ViteTarballPluginOptions {
  dest?: string
}
