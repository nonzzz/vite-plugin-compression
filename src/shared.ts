import fsp from 'fs/promises'
import path from 'path'

export function len<T extends ArrayLike<unknown>>(source: T) {
  return source.length
}

// [path][base].ext
// [path] is replaced with the directories to the original asset, included trailing
// [base] is replaced with the base ([name] + [ext]) of the original asset (image.png)
export function replaceFileName(staticPath: string, rule: string | ((id: string) => string)) {
  const template = typeof rule === 'function' ? rule(staticPath) : rule
  const { dir, base } = path.parse(staticPath)
  const p = dir ? dir + '/' : ''
  return template.replace(/\[path\]/, p).replace(/\[base\]/, base)
}

export function slash(path: string) {
  const isExtendedLengthPath = /^\\\\\?\\/.test(path)
  if (isExtendedLengthPath) { return path }
  return path.replace(/\\/g, '/')
}

export async function readAll(entry: string) {
  const paths = await Promise.all((await fsp.readdir(entry)).map((dir) => path.join(entry, dir)))
  let pos = 0
  const result: string[] = []
  while (pos !== len(paths)) {
    const dir = paths[pos]
    const stat = await fsp.stat(dir)
    if (stat.isDirectory()) {
      const dirs = await fsp.readdir(dir)
      paths.push(...dirs.map((sub) => path.join(dir, sub)))
    }
    if (stat.isFile()) {
      result.push(dir)
    }
    pos++
  }
  return result
}

const encoder = new TextEncoder()

export function stringToBytes(b: string | Uint8Array) {
  return typeof b === 'string' ? encoder.encode(b) : b
}

export function noop() {}
