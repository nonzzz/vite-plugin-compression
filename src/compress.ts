import fs from 'fs'
import path from 'path'
import { createPack } from 'tar-mini'
import util from 'util'
import zlib from 'zlib'
import type { BrotliOptions, InputType, ZlibOptions, ZstdOptions } from 'zlib'

import type { Algorithm, AlgorithmFunction, CoreAlgorithm, UserCompressionOptions } from './interface'
import { slash, stringToBytes } from './shared'

export function resolveAlgorithm(algorithm: Algorithm): CoreAlgorithm {
  if (algorithm === 'gz') { return 'gzip' }
  if (algorithm === 'brotli' || algorithm === 'br') { return 'brotliCompress' }
  if (algorithm === 'zstd') { return 'zstandard' }
  return algorithm
}
// Note: we should verify zstd support
// It add at `v23.8.0` and `v22.15.0`
export function ensureAlgorithm(userAlgorithm: Algorithm) {
  const resolvedAlgorithm = resolveAlgorithm(userAlgorithm)
  if (resolvedAlgorithm === 'zstandard') {
    const [major, minor] = process.versions.node.split('.').map((s) => +s)
    const isSupported = (major > 23) ||
      (major === 23 && minor >= 8) ||
      (major === 22 && minor >= 15)
    if (!isSupported) {
      throw new Error(
        `Node.js ${process.versions.node} does not support zstd compression. ` +
          `Requires Node.js >= 22.15.0 or >= 23.8.0`
      )
    }
    if (!zlib.zstdCompress) {
      throw new Error('zstd compression is not available in this Node.js build')
    }
    return {
      algorithm: util.promisify(zlib.zstdCompress)
    }
  }
  const algorithm = resolvedAlgorithm in zlib ? resolvedAlgorithm : 'gzip'

  return {
    algorithm: util.promisify(zlib[algorithm])
  }
}

export async function compress<T extends UserCompressionOptions | undefined>(
  buf: InputType,
  compress: AlgorithmFunction<T>,
  options: T
) {
  try {
    const res = await compress(buf, options)
    return res
  } catch (error) {
    return Promise.reject(error as Error)
  }
}

export const defaultCompressionOptions: {
  [algorithm in CoreAlgorithm]: algorithm extends 'brotliCompress' ? BrotliOptions : algorithm extends 'zstd' ? ZstdOptions : ZlibOptions
} = {
  gzip: {
    level: zlib.constants.Z_BEST_COMPRESSION
  },
  brotliCompress: {
    params: {
      [zlib.constants.BROTLI_PARAM_QUALITY]: zlib.constants.BROTLI_MAX_QUALITY
    }
  },
  deflate: {
    level: zlib.constants.Z_BEST_COMPRESSION
  },
  deflateRaw: {
    level: zlib.constants.Z_BEST_COMPRESSION
  },
  // I don't know what the best default options for zstd are, so using an empty object
  zstandard: {}
}

interface TarballOptions {
  dests: string[]
  root: string
}

interface TarballFileMeta {
  filename: string
  content: string | Uint8Array
}

export function createTarBall() {
  const pack = createPack()

  const wss: fs.WriteStream[] = []

  const options: TarballOptions = {
    dests: [],
    root: ''
  }

  const add = (meta: TarballFileMeta) => {
    pack.add(stringToBytes(meta.content), { filename: meta.filename })
  }

  const setup = (tarballOPtions: TarballOptions) => {
    Object.assign(options, tarballOPtions)

    options.dests.forEach((dest) => {
      const expected = slash(path.resolve(options.root, dest + '.tar'))
      const parent = slash(path.dirname(expected))
      if (slash(options.root) !== parent) {
        fs.mkdirSync(parent, { recursive: true })
      }
      const w = fs.createWriteStream(expected)
      wss.push(w)
    })
  }

  const done = async () => {
    pack.done()
    await Promise.all(wss.map((w) =>
      new Promise<void>((resolve, reject) => {
        w.on('error', reject)
        w.on('finish', resolve)
        pack.receiver.pipe(w)
      })
    ))
    wss.length = 0
  }

  const context = {
    add,
    setup,
    done
  }

  return context
}
