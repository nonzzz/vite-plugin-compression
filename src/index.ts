import { createFilter } from '@rollup/pluginutils'
import ansis from 'ansis'
import fs from 'fs'
import fsp from 'fs/promises'
import os from 'os'
import path from 'path'
import type { Logger, Plugin, ResolvedConfig } from 'vite'
import { compress, createTarBall, defaultCompressionOptions, ensureAlgorithm, resolveAlgorithm } from './compress'
import type {
  Algorithm,
  AlgorithmFunction,
  AlgorithmToZlib,
  DefineAlgorithmResult,
  GenerateBundle,
  UserCompressionOptions,
  ViteCompressionPluginOption,
  ViteTarballPluginOptions
} from './interface'
import { captureViteLogger, len, noop, readAll, replaceFileName, slash, stringToBytes } from './shared'
import { createConcurrentQueue } from './task'

const VITE_INTERNAL_ANALYSIS_PLUGIN = 'vite:build-import-analysis'
const VITE_COMPRESSION_PLUGIN = 'vite-plugin-compression'
const VITE_COPY_PUBLIC_DIR = 'copyPublicDir'
const MAX_CONCURRENT = (() => {
  const cpus = os.cpus() || { length: 1 }
  if (cpus.length === 1) { return 10 }
  return Math.max(1, cpus.length - 1)
})()

interface CompressionPluginAPI {
  staticOutputs: Set<string>
  done: Promise<void>
}

function handleOutputOption(conf: ResolvedConfig) {
  // issue #39
  // In some case like vite-plugin-legacy will set an empty output item
  // we should skip it.

  // Using full path. I find if we using like `dist` or others path it can't
  // work on monorepo
  // eg:
  // yarn --cwd @pkg/website build
  const outputs: Set<string> = new Set()
  const prepareAbsPath = (root: string, sub: string) => slash(path.resolve(root, sub))
  if (conf.build.rollupOptions?.output) {
    const outputOptions = Array.isArray(conf.build.rollupOptions.output)
      ? conf.build.rollupOptions.output
      : [conf.build.rollupOptions.output]
    outputOptions.forEach((opt) => {
      if (typeof opt === 'object' && !len(Object.keys(opt))) { return }
      outputs.add(prepareAbsPath(conf.root, opt.dir || conf.build.outDir))
    })
  } else {
    outputs.add(prepareAbsPath(conf.root, conf.build.outDir))
  }
  return outputs
}

async function handleStaticFiles(config: ResolvedConfig, callback: (file: string, assets: string) => void) {
  const baseCondit = VITE_COPY_PUBLIC_DIR in config.build ? config.build.copyPublicDir : true
  if (config.publicDir && baseCondit && fs.existsSync(config.publicDir)) {
    const staticAssets = await readAll(config.publicDir)
    const publicPath = path.join(config.root, path.relative(config.root, config.publicDir))
    staticAssets.forEach((assets) => {
      const file = slash(path.relative(publicPath, assets))
      callback(file, assets)
    })
  }
}

function createEnhancedError(error: Error, context: string): Error {
  if (error.name !== 'AggregateError' || !('errors' in error)) {
    return error
  }

  const aggregateError = error as AggregateError
  const errorCount = aggregateError.errors.length

  const messages = aggregateError.errors.map((err: Error, index: number) => {
    const message = err.message || 'Unknown error'
    return `[${index + 1}] ${message}`
  })

  const stacks = aggregateError.errors.map((err: Error, index: number) => {
    const stack = err.stack || err.message || 'No stack trace available'
    return `--- Error ${index + 1} ---\n${stack}`
  })

  const enhancedMessage = [
    `Compression failed in ${context} with ${errorCount} error(s):`,
    '',
    ...messages
  ].join('\n')

  const enhancedError = new Error(enhancedMessage)
  enhancedError.name = 'CompressionError'

  enhancedError.stack = [
    `CompressionError: ${enhancedMessage}`,
    '',
    'Combined stack traces:',
    ...stacks
  ].join('\n')

  enhancedError.cause = aggregateError.errors

  return enhancedError
}

function tarball(opts: ViteTarballPluginOptions = {}): Plugin {
  const { dest: userDest } = opts
  const statics: string[] = []
  const outputs: string[] = []
  let dests: string[] = []
  let root = process.cwd()
  const tarball = createTarBall()
  const queue = createConcurrentQueue(MAX_CONCURRENT)
  let ctx: ReturnType<typeof compression.getPluginAPI>
  return {
    name: 'vite-plugin-tarball',
    enforce: 'post',
    async configResolved(config) {
      outputs.push(...handleOutputOption(config))
      root = config.root
      dests = userDest ? [userDest] : outputs
      // No need to add source to pack in configResolved stage
      // If we do at the start stage. The build task will be slow.
      ctx = compression.getPluginAPI(config.plugins)
      if (!ctx) {
        await handleStaticFiles(config, (file) => {
          statics.push(file)
        })
      }
      // create dest dir
      tarball.setup({ dests, root })
    },
    writeBundle(_, bundles) {
      for (const fileName in bundles) {
        const bundle = bundles[fileName]
        tarball.add({ filename: fileName, content: bundle.type === 'asset' ? bundle.source : bundle.code })
      }
    },
    async closeBundle() {
      if (ctx) {
        await ctx.done
      }
      if (!statics.length && ctx && ctx.staticOutputs.size) {
        statics.push(...ctx.staticOutputs)
      }

      for (const dest of outputs) {
        for (const file of statics) {
          queue.enqueue(async () => {
            const p = path.join(dest, file)
            const buf = await fsp.readFile(p)
            tarball.add({ filename: file, content: buf })
          })
        }
      }
      await queue.wait().catch((err: Error) => {
        this.error(createEnhancedError(err, 'tarball creation'))
      })
      await tarball.done()
    }
  }
}

function hijackGenerateBundle(plugin: Plugin, afterHook: GenerateBundle) {
  const hook = plugin.generateBundle
  if (typeof hook === 'object' && hook.handler) {
    const fn = hook.handler
    hook.handler = async function handler(this, ...args: Parameters<GenerateBundle>) {
      await fn.apply(this, args)
      await afterHook.apply(this, args)
    }
  }
  if (typeof hook === 'function') {
    plugin.generateBundle = async function handler(this, ...args: Parameters<GenerateBundle>) {
      await hook.apply(this, args)
      await afterHook.apply(this, args)
    }
  }
}

function compression(
  opts: ViteCompressionPluginOption = {}
): Plugin {
  const {
    include = /\.(html|xml|css|json|js|mjs|svg|yaml|yml|toml)$/,
    exclude,
    threshold = 0,
    algorithms: userAlgorithms = ['gzip', 'brotliCompress'],
    filename,
    deleteOriginalAssets = false,
    skipIfLargerOrEqual = true
  } = opts

  const algorithms: DefineAlgorithmResult[] = []

  userAlgorithms.forEach((algorithm) => {
    if (typeof algorithm === 'string') {
      algorithms.push(defineAlgorithm(algorithm))
    } else if (typeof algorithm === 'object' && Array.isArray(algorithm)) {
      algorithms.push(algorithm)
    }
  })

  const filter = createFilter(include, exclude)

  const statics: string[] = []
  const outputs: string[] = []

  // vite internal vite:reporter don't write any log info to stdout. So we only capture the built in message
  // and print static process to stdout. I don't want to complicate things. So about message aligned with internal
  // result. I never deal with it.
  const { msgs, cleanup } = captureViteLogger()

  let logger: Logger

  let root: string = process.cwd()

  const zlibs = algorithms.map(([algorithm, options]) => ({
    algorithm,
    algorithmFunction: typeof algorithm === 'string' ? ensureAlgorithm(algorithm).algorithm : algorithm,
    options,
    filename: filename ??
      (algorithm === 'brotliCompress' ? '[path][base].br' : algorithm === 'zstandard' ? '[path][base].zst' : '[path][base].gz')
  }))

  const queue = createConcurrentQueue(MAX_CONCURRENT)

  const generateBundle: GenerateBundle = async function handler(_, bundles) {
    for (const fileName in bundles) {
      if (!filter(fileName)) { continue }
      const bundle = bundles[fileName]
      const source = stringToBytes(bundle.type === 'asset' ? bundle.source : bundle.code)
      const size = len(source)
      if (size < threshold) { continue }
      queue.enqueue(async () => {
        for (let i = 0; i < zlibs.length; i++) {
          const z = zlibs[i]
          const flag = i === zlibs.length - 1
          const name = replaceFileName(fileName, z.filename, { options: z.options, algorithm: z.algorithm })
          const compressed = await compress(source, z.algorithmFunction, z.options)
          if (skipIfLargerOrEqual && len(compressed) >= size) { return }
          // #issue 30 31
          // https://rollupjs.org/plugin-development/#this-emitfile
          if (flag) {
            if (deleteOriginalAssets || fileName === name) { Reflect.deleteProperty(bundles, fileName) }
          }
          this.emitFile({ type: 'asset', fileName: name, source: compressed })
        }
      })
    }
    await queue.wait().catch((err: Error) => {
      this.error(createEnhancedError(err, 'bundle compression'))
    })
  }

  const doneResolver: { resolve: () => void } = { resolve: noop }

  const pluginContext: CompressionPluginAPI = {
    staticOutputs: new Set(),
    done: new Promise((resolve) => {
      doneResolver.resolve = resolve
    })
  }

  const numberFormatter = new Intl.NumberFormat('en', {
    maximumFractionDigits: 2,
    minimumFractionDigits: 2
  })

  const displaySize = (bytes: number) => {
    return `${numberFormatter.format(bytes / 1000)} kB`
  }

  const plugin = <Plugin> {
    name: VITE_COMPRESSION_PLUGIN,
    apply: 'build',
    enforce: 'post',
    api: pluginContext,
    async configResolved(config) {
      // hijack vite's internal `vite:build-import-analysis` plugin.So we won't need process us chunks at closeBundle anymore.
      // issue #26
      // https://github.com/vitejs/vite/blob/716286ef21f4d59786f21341a52a81ee5db58aba/packages/vite/src/node/build.ts#L566-L611
      // Vite follow rollup option as first and the configResolved Hook don't expose merged conf for user. :(
      // Someone who like using rollupOption. `config.build.outDir` will not as expected.
      outputs.push(...handleOutputOption(config))
      // Vite's pubic build: https://github.com/vitejs/vite/blob/HEAD/packages/vite/src/node/build.ts#L704-L709
      // copyPublicDir minimum version 3.2+
      // No need check size here.
      await handleStaticFiles(config, (file) => {
        statics.push(file)
      })
      const viteAnalyzerPlugin = config.plugins.find((p) => p.name === VITE_INTERNAL_ANALYSIS_PLUGIN)
      if (!viteAnalyzerPlugin) {
        throw new Error("[vite-plugin-compression] Can't be work in versions lower than vite at 2.0.0")
      }
      hijackGenerateBundle(viteAnalyzerPlugin, generateBundle)

      logger = config.logger

      root = config.root
    },
    async closeBundle() {
      const compressedMessages: Array<{ dest: string, file: string, size: number }> = []

      const compressAndHandleFile = async (filePath: string, file: string, dest: string) => {
        const buf = await fsp.readFile(filePath)

        for (let i = 0; i < zlibs.length; i++) {
          const z = zlibs[i]
          const flag = i === zlibs.length - 1
          const compressed = await compress(buf, z.algorithmFunction, z.options)
          if (skipIfLargerOrEqual && len(compressed) >= len(buf)) {
            if (!pluginContext.staticOutputs.has(file)) { pluginContext.staticOutputs.add(file) }
            return
          }

          const fileName = replaceFileName(file, z.filename, { options: z.options, algorithm: z.algorithm })
          if (!pluginContext.staticOutputs.has(fileName)) { pluginContext.staticOutputs.add(fileName) }

          const outputPath = path.join(dest, fileName)
          if (flag) {
            if (deleteOriginalAssets && outputPath !== filePath) {
              await fsp.rm(filePath, { recursive: true, force: true })
            }
          }
          await fsp.writeFile(outputPath, compressed)
          compressedMessages.push({ dest: path.relative(root, dest) + '/', file: fileName, size: len(compressed) })
        }
      }

      const processFile = async (dest: string, file: string) => {
        const filePath = path.join(dest, file)
        if (!filter(filePath) && !pluginContext.staticOutputs.has(file)) {
          pluginContext.staticOutputs.add(file)
          return
        }
        const { size } = await fsp.stat(filePath)
        if (size < threshold) {
          if (!pluginContext.staticOutputs.has(file)) {
            pluginContext.staticOutputs.add(file)
          }
          return
        }
        await compressAndHandleFile(filePath, file, dest)
      }

      // parallel run
      for (const dest of outputs) {
        for (const file of statics) {
          queue.enqueue(() => processFile(dest, file))
        }
      }

      // issue #18
      // In somecase. Like vuepress it will called vite build with `Promise.all`. But it's concurrency. when we record the
      // file fd. It had been changed. So that we should catch the error
      await queue.wait().catch((e: unknown) => e)
      doneResolver.resolve()
      cleanup()

      if (logger) {
        const paddingSize = compressedMessages.reduce((acc, cur) => {
          const full = cur.dest + cur.file
          return Math.max(acc, full.length)
        }, 0)

        for (const { dest, file, size } of compressedMessages) {
          const paddedFile = file.padEnd(paddingSize)
          logger.info(ansis.dim(dest) + ansis.green(paddedFile) + ansis.bold(ansis.dim(displaySize(size))))
        }
      }

      for (const msg of msgs) {
        console.info(msg)
      }
    }
  }

  return plugin
}

compression.getPluginAPI = (plugins: readonly Plugin[]): CompressionPluginAPI | undefined =>
  (plugins.find((p) => p.name === VITE_COMPRESSION_PLUGIN) as Plugin<CompressionPluginAPI>)?.api

export function defineAlgorithm<T extends Algorithm | UserCompressionOptions | AlgorithmFunction<UserCompressionOptions>>(
  algorithm: T extends Algorithm | AlgorithmFunction<UserCompressionOptions> ? T : AlgorithmFunction<Exclude<T, string>>,
  options?: T extends Algorithm ? AlgorithmToZlib[T] : T extends AlgorithmFunction<UserCompressionOptions> ? UserCompressionOptions : T
): DefineAlgorithmResult<T extends Algorithm | AlgorithmFunction<UserCompressionOptions> ? UserCompressionOptions : T> {
  if (typeof algorithm === 'string') {
    const resolvedAlgorithm = resolveAlgorithm(algorithm)
    if (resolvedAlgorithm in defaultCompressionOptions) {
      const opts = { ...defaultCompressionOptions[resolvedAlgorithm] }
      if (options) {
        Object.assign(opts, options)
      }
      return [resolvedAlgorithm, opts]
    }
    throw new Error(`[vite-plugin-compression] Unsupported algorithm: ${algorithm}`)
  }
  // @ts-expect-error no need to check
  return [algorithm, options || {}]
}

export { compression, tarball }

export default compression

export type { Algorithm, CompressionOptions, ViteCompressionPluginOption, ViteTarballPluginOptions } from './interface'
