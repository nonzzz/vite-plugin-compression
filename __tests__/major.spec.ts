import fs from 'fs'
import path from 'path'
import util from 'util'
import { afterAll, assert, describe, expect, it } from 'vitest'
import zlib from 'zlib'
import { compression, defineAlgorithm } from '../src'
import { readAll } from '../src/shared'
import { createDisk, getId, mockBuild } from './shared/kit.mjs'

describe('compression plugin', () => {
  const { root, destroy, dir } = createDisk('compression-plugin')
  afterAll(destroy)

  it('include only', async () => {
    const { output } = await mockBuild('normal', root, { plugins: [compression({ include: /\.(js)$/ })] })
    expect((await readAll(output)).filter((s) => s.endsWith('.gz')).length).toBe(1)
  })

  it('exclude only', async () => {
    const { output } = await mockBuild('normal', root, {
      plugins: [compression({ exclude: /\.(html)$/, skipIfLargerOrEqual: false })]
    })
    expect((await readAll(output)).filter((s) => s.endsWith('.gz')).length).toBe(2)
  })

  it('threshold', async () => {
    const { output } = await mockBuild('normal', root, { plugins: [compression({ threshold: 100 })] })
    expect((await readAll(output)).filter((s) => s.endsWith('.gz')).length).toBe(2)
  })

  it('algorithm', async () => {
    const { output } = await mockBuild('normal', root, {
      plugins: [compression({ algorithms: ['gzip'], skipIfLargerOrEqual: false })]
    })
    expect((await readAll(output)).filter((s) => s.endsWith('.gz')).length).toBe(3)
  })

  it('custom alorithm', async () => {
    const { output } = await mockBuild('normal', root, {
      plugins: [compression({
        algorithms: [defineAlgorithm((buf, opt) => {
          return util.promisify(zlib.gzip)(buf, opt)
        }, { level: 9 })],
        skipIfLargerOrEqual: false
      })]
    })
    expect((await readAll(output)).filter((s) => s.endsWith('.gz')).length).toBe(3)
  })

  it('brotliCompress', async () => {
    const { output } = await mockBuild('normal', root, {
      plugins: [compression({
        algorithms: ['brotliCompress'],
        skipIfLargerOrEqual: false
      })]
    })
    expect((await readAll(output)).filter((s) => s.endsWith('.br')).length).toBe(3)
  })
  it('delete original assets', async () => {
    const { output } = await mockBuild('normal', root, {
      plugins: [compression({
        algorithms: ['gzip', 'brotliCompress'],
        deleteOriginalAssets: true,
        skipIfLargerOrEqual: false
      })]
    })
    expect((await readAll(output)).length).toBe(6)
  })

  it('filename', async () => {
    const { output } = await mockBuild('normal', root, {
      plugins: [compression({
        skipIfLargerOrEqual: false,
        algorithms: ['gzip'],
        filename: 'fake/[base].gz'
      })]
    })
    const result = await readAll(path.join(output, 'fake'))
    expect(result.filter((s) => s.endsWith('.gz')).length).toBe(3)
  })

  it('multiple', async () => {
    const { output } = await mockBuild('normal', root, {
      plugins: [
        compression({ skipIfLargerOrEqual: false, include: /\.(css)$/ })
      ]
    })
    const result = await readAll(output)
    const gzip = result.filter((s) => s.endsWith('.gz'))
    const br = result.filter((s) => s.endsWith('.br'))
    expect(gzip.length).toBe(br.length)
  })

  it('dynamic import source', async () => {
    const { output } = await mockBuild('dynamic', root, {
      plugins: [compression({
        skipIfLargerOrEqual: false,
        deleteOriginalAssets: true
      })]
    })
    expect((await readAll(output)).filter((s) => s.endsWith('.gz')).length).toBe(4)
  })

  describe('Public assets', () => {
    it('normal', async () => {
      const { output } = await mockBuild('public-assets', root, {
        plugins: [compression({
          skipIfLargerOrEqual: false,
          deleteOriginalAssets: true,
          exclude: /\.(html)$/
        })]
      })
      expect((await readAll(output)).filter((s) => s.endsWith('.gz')).length).toBe(3)
    })
    it('nesting', async () => {
      const { output } = await mockBuild('public-assets-nest', root, {
        plugins: [compression({
          skipIfLargerOrEqual: false,
          deleteOriginalAssets: true,
          exclude: /\.(html)$/
        })]
      })
      const result = await readAll(output)
      expect(result.filter((s) => s.endsWith('.gz')).length).toBe(6)
    })

    it('threshold', async () => {
      const { output } = await mockBuild('public-assets-nest', root, {
        plugins: [compression({
          skipIfLargerOrEqual: false,
          deleteOriginalAssets: true,
          exclude: /\.(html)$/,
          threshold: 1024 * 2
        })]
      })
      const result = await readAll(output)
      expect(result.filter((s) => s.endsWith('.gz')).length).toBe(0)
    })
  })

  it('amazon s3', async () => {
    const { output } = await mockBuild('dynamic', root, {
      plugins: [compression({
        skipIfLargerOrEqual: false,
        algorithms: ['gzip'],
        filename: '[path][base]',
        deleteOriginalAssets: true
      })]
    })
    const result = await readAll(output)
    const cssFiles = result.filter((v) => v.endsWith('.css'))
    assert(cssFiles.length === 1)
    expect(zlib.unzipSync(fs.readFileSync(cssFiles[0])).toString()).toBe(
      '.pr{padding-right:30px}.pl{padding-left:30px}.mt{margin-top:30px}\n'
    )
  })

  describe('rollup options', () => {
    it('multiple outputs', async () => {
      const another = path.join(dir, getId())
      const another2 = path.join(dir, getId())
      await mockBuild('public-assets-nest', root, {
        plugins: [compression({ skipIfLargerOrEqual: false, exclude: /\.(html)$/ })],
        build: {
          rollupOptions: {
            output: [{ dir: another }, { dir: another2 }]
          }
        }
      })
      const result1 = await readAll(another)
      const result2 = await readAll(another2)
      assert(result1.length === result2.length)
    })
  })
  it('zstd version v23.8.0 and v22.15.0', async () => {
    const p = () => compression({ include: /\.(js)$/, algorithms: ['zstd'] })
    const [major, minor] = process.versions.node.split('.').map((s) => +s)
    if ((+major === 23 && +minor < 8) || (+major < 22 || (+major === 22 && +minor < 15))) {
      expect(p).toThrowError(
        `Node.js ${process.versions.node} does not support zstd compression. ` +
          `Requires Node.js >= 22.15.0 or >= 23.8.0`
      )
    } else {
      expect(defineAlgorithm('zstd')).toHaveLength(2)
      const { output } = await mockBuild('normal', root, { plugins: [p()] })
      expect((await readAll(output)).filter((s) => s.endsWith('.zst')).length).toBe(1)
    }
  })
  it('pretty error stack', async () => {
    await expect(async () =>
      mockBuild('normal', root, {
        plugins: [compression({
          algorithms: ['gz'],
          filename: () => {
            throw new Error('Test error for pretty stack')
          }
        })]
      })
    ).rejects.toThrowErrorMatchingSnapshot()
  })
})
