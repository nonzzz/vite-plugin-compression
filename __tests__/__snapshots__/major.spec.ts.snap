// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`compression plugin > pretty error stack 1`] = `
[CompressionError: [31m[vite:build-import-analysis] Compression failed in bundle compression with 3 error(s):

[1] Test error for pretty stack
[2] Test error for pretty stack
[3] Test error for pretty stack[39m]
`;
